onnxruntime/wasm/api.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "api.h"
+
+#include "core/session/onnxruntime_cxx_api.h"
+
+#include <vector>
+
+namespace {
+Ort::Env* g_env;
+}  // namespace
+
+void ort_init() {
+  OrtLoggingLevel logging_level = ORT_LOGGING_LEVEL_WARNING;
+  g_env = new Ort::Env{logging_level, "Default"};
+}
+
+Ort::Session* ort_create_session(void* data, size_t data_length) {
+  Ort::SessionOptions session_options;
+  session_options.SetLogId("onnxruntime");
+
+  // disable thread pool for now since not all major browsers support WebAssembly threading.
+  session_options.SetIntraOpNumThreads(1);
+
+  return new Ort::Session(*g_env, data, data_length, session_options);
+}
+
+void ort_release_session(Ort::Session* session) {
+  delete session;
+}
+
+size_t ort_get_input_count(Ort::Session* session) {
+  return session->GetInputCount();
+}
+
+size_t ort_get_output_count(Ort::Session* session) {
+  return session->GetOutputCount();
+}
+
+char* ort_get_input_name(Ort::Session* session, size_t index) {
+  Ort::AllocatorWithDefaultOptions allocator;
+  return session->GetInputName(index, allocator);
+}
+
+char* ort_get_output_name(Ort::Session* session, size_t index) {
+  Ort::AllocatorWithDefaultOptions allocator;
+  return session->GetOutputName(index, allocator);
+}
+
+void ort_free(void* ptr) {
+  Ort::AllocatorWithDefaultOptions allocator;
+  allocator.Free(ptr);
+}
+
+OrtValue* ort_create_tensor(int data_type, void* data, size_t data_length, size_t* dims, size_t dims_length) {
+  std::vector<int64_t> shapes(dims_length);
+  for (size_t i = 0; i < dims_length; i++) {
+    shapes[i] = dims[i];
+  }
+
+  return Ort::Value::CreateTensor({}, data, data_length, dims_length > 0 ? shapes.data() : nullptr, dims_length, static_cast<ONNXTensorElementDataType>(data_type))
+      .release();
+}
+
+void ort_get_tensor_data(OrtValue* tensor, int* data_type, void** data, size_t** dims, size_t* dims_length) {
+  Ort::Value v{tensor};
+  auto info = v.GetTensorTypeAndShapeInfo();
+  size_t dims_len = info.GetDimensionsCount();
+  Ort::AllocatorWithDefaultOptions allocator;
+  size_t* p_dims = reinterpret_cast<size_t*>(allocator.Alloc(sizeof(size_t) * dims_len));
+  *data = v.GetTensorMutableData<void>();
+  *data_type = info.GetElementType();
+  *dims_length = dims_len;
+  auto shape = info.GetShape();
+  for (size_t i = 0; i < dims_len; i++) {
+    p_dims[i] = static_cast<size_t>(shape[i]);
+  }
+  *dims = p_dims;
+  v.release();
+}
+
+void ort_release_tensor(OrtValue* tensor) {
+  Ort::OrtRelease(tensor);
+}
+
+void ort_run(ort_session_handle_t session, const char** input_names, const ort_tensor_handle_t* inputs, size_t input_count, const char** output_names, size_t output_count, ort_tensor_handle_t* outputs) {
+  Ort::ThrowOnError(Ort::GetApi().Run(*session, Ort::RunOptions{nullptr}, input_names, inputs, input_count, output_names, output_count, outputs));
```

Comment:
>  Ort::ThrowOnError(Ort::GetApi().Run(*session, Ort::RunOptions{nullptr}, input_names, inputs, input_count, output_names, output_count,  [](start = 0, length = 136)

Is it correct that it's a C API, but you're throwing an exception here? That seems slightly off.

Reply:
It's exposed as C API but it is not. Those C-Style functions are not consumed by any C++ users; instead it is consumed by emcc to generate the .wasm file, which will be used by Javascript code. With some build flags exceptions can be caught.

Reply:
I see this build flag onnxruntime_ENABLE_WEBASSEMBLY_EXCEPTION_CATCHING which could be on or off. If it's off, should this still throw or do something else?

---
In reply to: [598992672](https://github.com/microsoft/onnxruntime/pull/6478#discussion_r598992672) [](ancestors = 598992672)

Reply:
this flag only controls exception catching; it does not control exception throwing. to disable exception throwing we need `--disable_exceptions` (currently this flag can only be set in minimal build).

more details: https://github.com/emscripten-core/emscripten/blob/main/src/settings.js#L637

Reply:
What's the use case to allow exceptions, but disable catching them?

Reply:
Yeah - this still seems like an open question. What happens if onnxruntime_ENABLE_WEBASSEMBLY_EXCEPTION_CATCHING is off and you throw? Who catches that?

Reply:
Emscripten's actual behavior is to generate code to output a string "exception thrown: <memory address>" to stdout and call `abort()` when an exception is thrown.

In other words, there is no benefits to allow exception throwing while at the same time disable exception catching. The only reason why we do this is because we cannot build ORT without exception throwing (unless minimal build).

Reply:
You're making a choice to call ThrowOnError aren't you? Couldn't we return an error status instead? An abort() is really hard to debug, so ideally you can avoid that.

---
In reply to: [605999138](https://github.com/microsoft/onnxruntime/pull/6478#discussion_r605999138) [](ancestors = 605999138)

Reply:
this is a good point. I update implementation of `OrtRun()` so now it returns `int` (actually `OrtErrorCode`).

I also create a backlog for removing `ThrowOnError` in all WASM API implementation (it uses ORT C++ API so it's used a lot) and I am trying to complete it by v1.8